beebop/visualise.py

```diff
@@ -160,14 +160,78 @@ def visualise_per_cluster(
 
     cluster_no = get_cluster_num(assign_cluster)
     output_folder = fs.output_visualisations(p_hash, cluster_no)
-    if external_to_poppunk_clusters:
-        internal_cluster = external_to_poppunk_clusters[assign_cluster]
-    else:
-        internal_cluster = assign_cluster
+    internal_cluster = get_internal_cluster(
+        external_to_poppunk_clusters,
+        assign_cluster,
+        p_hash,
+        fs,
+    )
 
-    wrapper.create_visualisations(cluster_no, internal_cluster)
+    wrapper.create_visualisations(
+        cluster_no, fs.include_file(p_hash, internal_cluster)
+    )
 
     replace_filehashes(output_folder, name_mapping)
     create_subgraph(output_folder, name_mapping, cluster_no)
     if is_last_cluster_to_process:
         os.remove(fs.tmp_output_metadata(p_hash))
+
+
+def get_internal_cluster(
+    external_to_poppunk_clusters: dict[str, list[str]],
+    assign_cluster: str,
+    p_hash: str,
+    fs: PoppunkFileStore,
+) -> str:
+    """
+    [Determines the internal cluster name based on
+    the external cluster mapping.
+    In the edge case where the external cluster maps
+    to multiple internal clusters,
+    it creates a combined include file.]
+
+    :param external_to_poppunk_clusters: [dict mapping external clusters
+        to internal clusters]
+    :param assign_cluster: [the external cluster number]
+    :param p_hash: [project hash to find input data (output from
+        assignClusters)]
+    :param fs: [PoppunkFileStore with paths to input data]
+    :return: [internal cluster name or combined include file name]
+    """
+    if not external_to_poppunk_clusters:
+        return assign_cluster
+
+    internal_clusters = external_to_poppunk_clusters[assign_cluster]
+    if len(internal_clusters) == 1:
+        return internal_clusters[0]
+
+    return create_combined_include_file(fs, p_hash, internal_clusters)
+
+
+def create_combined_include_file(
+    fs: PoppunkFileStore,
+    p_hash: str,
+    internal_clusters: list[str],
+) -> str:
+    """
+    [Creates a combined include file for multiple internal clusters.
+    The combined file contains the contents of
+    all specified internal clusters.]
+
+    :param fs: [PoppunkFileStore with paths to input data]
+    :param p_hash: [project hash to find input data (output from
+        assignClusters)]
+    :param internal_clusters: [list of internal cluster names to combine]
+    :return: [combined internal cluster]
+    """
+    combined_internal_cluster = "_".join(internal_clusters)
+    with open(
+        fs.include_file(p_hash, combined_internal_cluster), "w"
```

Comment:
So the include files for all the individual internal clusters already exist, and this just writes out a combined file as required.?

Reply:
yup the include files are created via the assign step. then the visualise step uses them